Jenkinsfile

```diff
@@ -0,0 +1,164 @@
+def registry = '562955126301.dkr.ecr.eu-west-2.amazonaws.com'
+def regCredsId = 'ecr:eu-west-2:ecr-user'
+def kubeCredsId = 'awskubeconfig002'
+def ingressServer = "ffc.aws-int.defra.cloud"
+def imageName = 'ffc-demo-user-service'
+def repoName = 'ffc-demo-user-service'
+def repoUrl = ''
+def commitSha = ''
+def branch = ''
+def pr = ''
+def mergedPrNo = ''
+def containerTag = ''
+
+def getMergedPrNo() {
+    def mergedPrNo = sh(returnStdout: true, script: "git log --pretty=oneline --abbrev-commit -1 | sed -n 's/.*(#\\([0-9]\\+\\)).*/\\1/p'").trim()
+    return mergedPrNo ? "pr$mergedPrNo" : ''
+}
+
+def getRepoURL() {
+  return sh(returnStdout: true, script: "git config --get remote.origin.url").trim()
+}
+
+def getCommitSha() {
+  return sh(returnStdout: true, script: "git rev-parse HEAD").trim()
+}
+
+def getVariables(repoName) {
+    def branch = BRANCH_NAME
+    // use the git API to get the open PR for a branch
+    // Note: This will cause issues if one branch has two open PRs
+    def pr = sh(returnStdout: true, script: "curl https://api.github.com/repos/DEFRA/$repoName/pulls?state=open | jq '.[] | select(.head.ref == \"$branch\") | .number'").trim()
+    def rawTag = pr == '' ? branch : "pr$pr"
+    def containerTag = rawTag.replaceAll(/[^a-zA-Z0-9]/, '-').toLowerCase()
+    return [branch, pr, containerTag,  getMergedPrNo(), getRepoURL(), getCommitSha()]
+}
+
+def updateGithubCommitStatus(message, state, repoUrl, commitSha) {
+  step([
+    $class: 'GitHubCommitStatusSetter',
+    reposSource: [$class: "ManuallyEnteredRepositorySource", url: repoUrl],
+    commitShaSource: [$class: "ManuallyEnteredShaSource", sha: commitSha],
+    errorHandlers: [[$class: 'ShallowAnyErrorHandler']],
+    statusResultSource: [ $class: "ConditionalStatusResultSource", results: [[$class: "AnyBuildResult", message: message, state: state]] ]
+  ])
+}
+
+def buildTestImage(name, suffix) {
+  sh 'docker image prune -f || echo could not prune images'
+  // CAUTION: This project uses a single docker-compose file for tests.
+  sh "docker-compose -p $name-$suffix -f docker-compose.test.yaml build --no-cache"
+}
+
+def runTests(name, suffix) {
+  // CAUTION: This project uses a single docker-compose file for tests.
+  try {
+    sh 'mkdir -p test-output'
+    sh 'chmod 777 test-output'
+    sh "docker-compose -p $name-$suffix -f docker-compose.test.yaml run ffc-demo-user-test"
```

Comment:
I think we should use a variable for the test service name and pass it to this function as the `$name` argument. For an example, see the API Gateway [Jenkinsfile](https://github.com/DEFRA/ffc-demo-api-gateway/blob/master/Jenkinsfile). This will help the later abstraction of Jenkinsfile functions into a library.

Also, note that currently the deploy step uses `$imageName` (i.e. `ffc-demo-user-service`) which is not built by the build step because we use only `docker-compose.test.yaml` for tests. Again, the API Gateway example shows the changes required to build two images (Docker cache will make the second build really fast since they're from the same Dockerfile and context) so we can deploy the `-service` image after testing with the `-test` image.

Reply:
That's a fair point, but I think it's reasonable to ensure the functions match across all the repositories to make migrating to the shared library easier.

I've just noticed that the `pushContainerImage` function actually builds the "production" image first so the extra build step I added to the API Gateway was redundant. I'll take that one out so it matches the others more closely.

I still feel that we should use a `testImageName` variable here so that the `runTests` function doesn't diverge from other Jenkinsfiles. It just needs declaring at the top and passing instead of `imageName` into `buildTestImage` and `runTests`.

Reply:
The additional `testImageName` parameter, which would actually be the service name, isn't required in the implementation demonstrated in the ffc-demo-web project as the service name is the same for the tests and the main service.

There's a chance the method signature won't have to change at all when migrated to the library so it seems reasonable to leave it as is.

As a side note, consistent naming of services and images would remove the need for an additional testImageName parameter, i.e. `"$imageName-test"`

The comment `there should be a 'build production image' step here`, is erroneous, as it does build one before the push to the registry.

Reply:
True, the method signature is the same. Fine to leave as-is if you prefer @Matthew-Collins-Defra.

The comment about building a production image is consistent with other pipelines that have been merged, but does seem a bit pointless. I don't mind it staying until the Dockerfile is rewritten to actually build a production image.

Reply:
All good points, but best dealt with later, when we've resolved how tests are going to be run (single or multiple docker-compose) and this has been pulled out into a library of Jenkins stuff.